### PR TITLE
Refactor the type def model for locally defined record types

### LINKED ISSUE #20729: Refactor how locally defined anonymous record types are modeled
Currently for anonymous record types (and objects and errors too), the type node is wrapped in a type def and added as a top level node when building the AST. The downside to this approach is even the locally defined types are defined in the module scope. Because of this, you cannot refer to local variables when giving default values for record fields; it will complain about an undefined symbol. Need to fix this scoping issue in order to allow referring to local symbols from a locally defined record type descriptor.

### PR DESCRIPTION
**Pull request by pubudu91**
Fixes #20729

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 1]
```diff
@@ -460,6 +468,14 @@ void addBuiltInReferenceType(DiagnosticPos pos, Set<Whitespace> ws, String typeN
         addType(refType);
     }
 
+    void startErrorType() {
+        this.isInErrorType++;
+    }
+
+    void endErrorType() {
```
Comment (anupama-pathirage) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Can we change this to private method?

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: This is called from the `BLangParserListener`. So cannot make it private

Comment (anupama-pathirage) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: I don't see any usage of `endErrorType()` from the `BLangParserListener`. Only the `startErrorType()` is used there.

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Sorry, my bad. Yes, you are correct, can make it private.

**Comment by pubudu91** [comment]
Closing this for now. Need to see if this can be refactored in a better way, without introducing a cyclic dependency between symbol enter and symbol resolver.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 2]
```diff
@@ -408,12 +410,18 @@ private BLangRecordTypeNode populateRecordTypeNode(DiagnosticPos pos, Set<Whites
...
+        recordTypeNode.isLocal = isInLocalDefinition();
         this.varListStack.pop().forEach(variableNode -> {
             recordTypeNode.addField((SimpleVariableNode) variableNode);
         });
         return recordTypeNode;
     }
 
+    private boolean isInLocalDefinition() {
+        // TODO: When supporting local defs for errors, need to get rid of the second condition
+        return !this.blockNodeStack.isEmpty() && this.isInErrorType <= 0;
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Please create an issue for this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -719,9 +751,79 @@ public void visit(BLangRecordTypeNode recordTypeNode) {
...
+    public void visit(BLangUnionTypeNode unionTypeNode) {
+        List<BLangType> rewrittenMembers = new ArrayList<>();
+        unionTypeNode.memberTypeNodes.forEach(typeNode -> rewrittenMembers.add(rewrite(typeNode, env)));
+        unionTypeNode.memberTypeNodes = rewrittenMembers;
+        result = unionTypeNode;
+    }
+
+    @Override
+    public void visit(BLangErrorType errorType) {
+        errorType.detailType = rewrite(errorType.detailType, env);
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Add reason field as well.

**Comment by hasithaa** [comment]
Shall we add test cases to cover new scenarios?

**Pull request merged**